File: go/models/group_icons.go
package models

// GroupIcon describes a single entry in the curated set of emojis users can
// pick for a location group. Exported for validation, tests, and migrations;
// the frontend ships an equivalent list under
// frontend/src/constants/groupIcons.ts — both must be kept in sync.
type GroupIcon struct {
	Emoji    string
	Label    string
	Category string
}

// Category labels used by the frontend picker. Kept string-keyed so adding a
// new category only requires editing this file and its frontend sibling.
const (
	GroupIconCategoryHome    = "home"
	GroupIconCategoryWork    = "work"
	GroupIconCategoryStorage = "storage"
	GroupIconCategoryNature  = "nature"
	GroupIconCategoryTravel  = "travel"
	GroupIconCategoryHobbies = "hobbies"
	GroupIconCategoryMisc    = "misc"
)

// ValidGroupIcons is the authoritative list of acceptable group-icon values.
// An empty string is also accepted (icon is optional) — the emptiness check
// lives in the validator, not here, so callers iterating this slice see only
// real icons.
//
// Order within a category drives the order in which the frontend picker shows
// them; keep related icons adjacent. Changing an emoji here is a
// backward-incompatible change — drop a stale entry from the backend set,
// ship a data migration that clears the stale value from location_groups.icon,
// then remove it from the frontend.
var ValidGroupIcons = []GroupIcon{
	// Home & family
	{Emoji: "🏠", Label: "House", Category: GroupIconCategoryHome},
	{Emoji: "🏡", Label: "Home with garden", Category: GroupIconCategoryHome},
	{Emoji: "🏘️", Label: "Houses", Category: GroupIconCategoryHome},
	{Emoji: "🏰", Label: "Castle", Category: GroupIconCategoryHome},
	{Emoji: "🛏️", Label: "Bed", Category: GroupIconCategoryHome},
	{Emoji: "🛋️", Label: "Couch", Category: GroupIconCategoryHome},

	// Work & office
	{Emoji: "🏢", Label: "Office", Category: GroupIconCategoryWork},
	{Emoji: "🏛️", Label: "Institution", Category: GroupIconCategoryWork},
	{Emoji: "💼", Label: "Briefcase", Category: GroupIconCategoryWork},
	{Emoji: "💻", Label: "Laptop", Category: GroupIconCategoryWork},
	{Emoji: "🖨️", Label: "Printer", Category: GroupIconCategoryWork},
	{Emoji: "📇", Label: "Card index", Category: GroupIconCategoryWork},

	// Storage
	{Emoji: "📦", Label: "Box", Category: GroupIconCategoryStorage},
	{Emoji: "🗄️", Label: "Filing cabinet", Category: GroupIconCategoryStorage},
	{Emoji: "🗃️", Label: "Card file", Category: GroupIconCategoryStorage},
	{Emoji: "🗂️", Label: "Dividers", Category: GroupIconCategoryStorage},
	{Emoji: "📁", Label: "Folder", Category: GroupIconCategoryStorage},
	{Emoji: "🧰", Label: "Toolbox", Category: GroupIconCategoryStorage},

	// Nature
	{Emoji: "🌳", Label: "Tree", Category: GroupIconCategoryNature},
	{Emoji: "🌲", Label: "Evergreen", Category: GroupIconCategoryNature},
	{Emoji: "🌵", Label: "Cactus", Category: GroupIconCategoryNature},
	{Emoji: "🌷", Label: "Tulip", Category: GroupIconCategoryNature},
	{Emoji: "🍀", Label: "Four-leaf clover", Category: GroupIconCategoryNature},
	{Emoji: "⛰️", Label: "Mountain", Category: GroupIconCategoryNature},

	// Travel & transport
	{Emoji: "🚗", Label: "Car", Category: GroupIconCategoryTravel},
	{Emoji: "🚚", Label: "Truck", Category: GroupIconCategoryTravel},
	{Emoji: "🚲", Label: "Bicycle", Category: GroupIconCategoryTravel},
	{Emoji: "✈️", Label: "Plane", Category: GroupIconCategoryTravel},
	{Emoji: "🧳", Label: "Luggage", Category: GroupIconCategoryTravel},
	{Emoji: "⛵", Label: "Sailboat", Category: GroupIconCategoryTravel},

	// Hobbies
	{Emoji: "🎨", Label: "Art", Category: GroupIconCategoryHobbies},
	{Emoji: "🎸", Label: "Guitar", Category: GroupIconCategoryHobbies},
	{Emoji: "🎮", Label: "Video game", Category: GroupIconCategoryHobbies},
	{Emoji: "📚", Label: "Books", Category: GroupIconCategoryHobbies},
	{Emoji: "📷", Label: "Camera", Category: GroupIconCategoryHobbies},
	{Emoji: "⚽", Label: "Football", Category: GroupIconCategoryHobbies},
	{Emoji: "🎲", Label: "Dice", Category: GroupIconCategoryHobbies},
	{Emoji: "🧩", Label: "Jigsaw", Category: GroupIconCategoryHobbies},

	// Misc
	{Emoji: "⭐", Label: "Star", Category: GroupIconCategoryMisc},
	{Emoji: "💎", Label: "Gem", Category: GroupIconCategoryMisc},
	{Emoji: "🔥", Label: "Fire", Category: GroupIconCategoryMisc},
	{Emoji: "⚙️", Label: "Gear", Category: GroupIconCategoryMisc},
	{Emoji: "🛠️", Label: "Tools", Category: GroupIconCategoryMisc},
	{Emoji: "🎯", Label: "Target", Category: GroupIconCategoryMisc},
	{Emoji: "🏷️", Label: "Label", Category: GroupIconCategoryMisc},
	{Emoji: "💡", Label: "Lightbulb", Category: GroupIconCategoryMisc},
}

// validGroupIconSet is a lookup table built from ValidGroupIcons once, at
// package init, so per-request validation stays O(1). Not exported — callers
// that need membership should use IsValidGroupIcon.
var validGroupIconSet = func() map[string]struct{} {
	set := make(map[string]struct{}, len(ValidGroupIcons))
	for _, ic := range ValidGroupIcons {
		set[ic.Emoji] = struct{}{}
	}
	return set
}()

// IsValidGroupIcon reports whether the given string is one of the curated
// group icons. An empty string is NOT considered valid here — the empty /
// unset case must be handled separately by the caller (e.g. the request
// validator allows empty, storage accepts empty). This keeps the function
// useful for both validation (reject nonsense) and data-migration filtering
// (rewrite anything-not-in-the-set to empty).
func IsValidGroupIcon(icon string) bool {
	if icon == "" {
		return false
	}
	_, ok := validGroupIconSet[icon]
	return ok
}

File: frontend/src/constants/groupIcons.ts
/**
 * Curated set of emoji icons a user can pick for a location group.
 *
 * Mirror of go/models/group_icons.go — the backend rejects any value that
 * isn't in this list (plus empty) with a 422. Both sides must stay in sync:
 * adding an emoji is backward-compatible, but removing/renaming one requires
 * a data migration clearing the stale value from location_groups.icon.
 *
 * Order within a category drives the picker grid order; keep related icons
 * adjacent. Issue #1255.
 */

export const GROUP_ICON_CATEGORY_HOME = 'home'
export const GROUP_ICON_CATEGORY_WORK = 'work'
export const GROUP_ICON_CATEGORY_STORAGE = 'storage'
export const GROUP_ICON_CATEGORY_NATURE = 'nature'
export const GROUP_ICON_CATEGORY_TRAVEL = 'travel'
export const GROUP_ICON_CATEGORY_HOBBIES = 'hobbies'
export const GROUP_ICON_CATEGORY_MISC = 'misc'

export interface GroupIcon {
  emoji: string
  label: string
  category: string
}

export interface GroupIconCategory {
  id: string
  label: string
}

export const GROUP_ICON_CATEGORIES: GroupIconCategory[] = [
  { id: GROUP_ICON_CATEGORY_HOME, label: 'Home' },
  { id: GROUP_ICON_CATEGORY_WORK, label: 'Work' },
  { id: GROUP_ICON_CATEGORY_STORAGE, label: 'Storage' },
  { id: GROUP_ICON_CATEGORY_NATURE, label: 'Nature' },
  { id: GROUP_ICON_CATEGORY_TRAVEL, label: 'Travel' },
  { id: GROUP_ICON_CATEGORY_HOBBIES, label: 'Hobbies' },
  { id: GROUP_ICON_CATEGORY_MISC, label: 'Misc' },
]

export const GROUP_ICONS: GroupIcon[] = [
  // Home & family
  { emoji: '🏠', label: 'House', category: GROUP_ICON_CATEGORY_HOME },
  { emoji: '🏡', label: 'Home with garden', category: GROUP_ICON_CATEGORY_HOME },
  { emoji: '🏘️', label: 'Houses', category: GROUP_ICON_CATEGORY_HOME },
  { emoji: '🏰', label: 'Castle', category: GROUP_ICON_CATEGORY_HOME },
  { emoji: '🛏️', label: 'Bed', category: GROUP_ICON_CATEGORY_HOME },
  { emoji: '🛋️', label: 'Couch', category: GROUP_ICON_CATEGORY_HOME },

  // Work & office
  { emoji: '🏢', label: 'Office', category: GROUP_ICON_CATEGORY_WORK },
  { emoji: '🏛️', label: 'Institution', category: GROUP_ICON_CATEGORY_WORK },
  { emoji: '💼', label: 'Briefcase', category: GROUP_ICON_CATEGORY_WORK },
  { emoji: '💻', label: 'Laptop', category: GROUP_ICON_CATEGORY_WORK },
  { emoji: '🖨️', label: 'Printer', category: GROUP_ICON_CATEGORY_WORK },
  { emoji: '📇', label: 'Card index', category: GROUP_ICON_CATEGORY_WORK },

  // Storage
  { emoji: '📦', label: 'Box', category: GROUP_ICON_CATEGORY_STORAGE },
  { emoji: '🗄️', label: 'Filing cabinet', category: GROUP_ICON_CATEGORY_STORAGE },
  { emoji: '🗃️', label: 'Card file', category: GROUP_ICON_CATEGORY_STORAGE },
  { emoji: '🗂️', label: 'Dividers', category: GROUP_ICON_CATEGORY_STORAGE },
  { emoji: '📁', label: 'Folder', category: GROUP_ICON_CATEGORY_STORAGE },
  { emoji: '🧰', label: 'Toolbox', category: GROUP_ICON_CATEGORY_STORAGE },

  // Nature
  { emoji: '🌳', label: 'Tree', category: GROUP_ICON_CATEGORY_NATURE },
  { emoji: '🌲', label: 'Evergreen', category: GROUP_ICON_CATEGORY_NATURE },
  { emoji: '🌵', label: 'Cactus', category: GROUP_ICON_CATEGORY_NATURE },
  { emoji: '🌷', label: 'Tulip', category: GROUP_ICON_CATEGORY_NATURE },
  { emoji: '🍀', label: 'Four-leaf clover', category: GROUP_ICON_CATEGORY_NATURE },
  { emoji: '⛰️', label: 'Mountain', category: GROUP_ICON_CATEGORY_NATURE },

  // Travel & transport
  { emoji: '🚗', label: 'Car', category: GROUP_ICON_CATEGORY_TRAVEL },
  { emoji: '🚚', label: 'Truck', category: GROUP_ICON_CATEGORY_TRAVEL },
  { emoji: '🚲', label: 'Bicycle', category: GROUP_ICON_CATEGORY_TRAVEL },
  { emoji: '✈️', label: 'Plane', category: GROUP_ICON_CATEGORY_TRAVEL },
  { emoji: '🧳', label: 'Luggage', category: GROUP_ICON_CATEGORY_TRAVEL },
  { emoji: '⛵', label: 'Sailboat', category: GROUP_ICON_CATEGORY_TRAVEL },

  // Hobbies
  { emoji: '🎨', label: 'Art', category: GROUP_ICON_CATEGORY_HOBBIES },
  { emoji: '🎸', label: 'Guitar', category: GROUP_ICON_CATEGORY_HOBBIES },
  { emoji: '🎮', label: 'Video game', category: GROUP_ICON_CATEGORY_HOBBIES },
  { emoji: '📚', label: 'Books', category: GROUP_ICON_CATEGORY_HOBBIES },
  { emoji: '📷', label: 'Camera', category: GROUP_ICON_CATEGORY_HOBBIES },
  { emoji: '⚽', label: 'Football', category: GROUP_ICON_CATEGORY_HOBBIES },
  { emoji: '🎲', label: 'Dice', category: GROUP_ICON_CATEGORY_HOBBIES },
  { emoji: '🧩', label: 'Jigsaw', category: GROUP_ICON_CATEGORY_HOBBIES },

  // Misc
  { emoji: '⭐', label: 'Star', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '💎', label: 'Gem', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '🔥', label: 'Fire', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '⚙️', label: 'Gear', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '🛠️', label: 'Tools', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '🎯', label: 'Target', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '🏷️', label: 'Label', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '💡', label: 'Lightbulb', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '🛡️', label: 'Shield', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '🔒', label: 'Lock', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '🧪', label: 'Lab', category: GROUP_ICON_CATEGORY_MISC },
  { emoji: '🧷', label: 'Safety pin', category: GROUP_ICON_CATEGORY_MISC },
]

const GROUP_ICON_SET = new Set(GROUP_ICONS.map((ic) => ic.emoji))

/**
 * Whether the given value is one of the curated group icons. An empty string
 * is NOT considered valid — callers that accept "unset" must handle that case
 * themselves.
 */
export function isValidGroupIcon(value: string): boolean {
  if (!value) return false
  return GROUP_ICON_SET.has(value)
}
